ev3dev2/motor.py

```diff
@@ -2182,6 +2189,138 @@ def turn_left(self, speed, degrees, brake=True, block=True):
+                left_previous = left_current
+                right_previous = right_current
+
+                # rotations = distance_mm/self.wheel.circumference_mm
+                left_rotations = float(left_ticks / self.left_motor.count_per_rot)
+                right_rotations = float(right_ticks / self.right_motor.count_per_rot)
+
+                # convert longs to floats and ticks to mm
+                left_mm = float(left_rotations * self.wheel.circumference_mm)
+                right_mm = float(right_rotations * self.wheel.circumference_mm)
+
+                # calculate distance we have traveled since last sampling
+                mm = (left_mm + right_mm) / 2.0
+
+                # accumulate total rotation around our center
+                self.theta += (left_mm - right_mm) / self.wheel_distance_mm
+
+                # and clip the rotation to plus or minus 360 degrees
+                self.theta -= float(int(self.theta/TWO_PI) * TWO_PI)
+
+                # now calculate and accumulate our position in mm
+                self.x_pos_mm += mm * math.sin(self.theta)
+                self.y_pos_mm += mm * math.cos(self.theta)
+
+                self.odometry_coordinates_log()
+                time.sleep(SLEEP_TIME)
+
+            self.odometry_thread_id = None
+
+        self.odometry_thread_run = True
+        self.odometry_thread_id = _thread.start_new_thread(_odometry_monitor, ())
+
+    def odometry_stop(self):
+        """
+        Signal the odometry thread to exit and wait for it to exit
+        """
+
+        if self.odometry_thread_id:
+            self.odometry_thread_run = False
+
+            while self.odometry_thread_id:
+                pass
+
+    def turn_to_angle(self, speed, angle_target_degrees, brake=True, block=True):
+
+        if angle_target_degrees < 0:
+            angle_target_degrees += 360
+
+        angle_current_degrees = math.degrees(self.theta)
+
+        if angle_current_degrees < 0:
+            angle_current_degrees += 360
+
+        angle_delta = angle_current_degrees - angle_target_degrees
+
+        log.info("%s: turn_to_angle %s, current angle %s, delta %s" %
+            (self, angle_target_degrees, angle_current_degrees, angle_delta))
+        self.odometry_coordinates_log()
+
+        # Do we rotate left or right and by how many degrees?
+        if angle_current_degrees > angle_target_degrees:
+            log.info("%s: turn right %s degrees" % (self, abs(angle_delta)))
+            self.turn_right(speed, angle_delta, brake, block)
+        else:
+            log.info("%s: turn left %s degrees" % (self, abs(angle_delta)))
+            self.turn_left(speed, angle_delta, brake, block)
+
+        self.odometry_coordinates_log()
+        log.info("\n\n\n")
+        # log.info("%s: angle is %.5f radians (%.5f degrees)." % (self, angle_rad, angle_target_degrees))
+
+    def on_to_coordinates(self, speed, x_target_mm, y_target_mm, brake=True, block=True):
```

Comment:
It would be cool if we had one which curved to put you at a point facing a direction... I have some algorithm ideas. That can wait though.

Reply:
yeah telling the robot to drive in a arc to reach the destination is very doable (we have arc_left and arc_right already).